### PR TITLE
[board] support of the Chimera autopilot based on STM32F7

### PR DESCRIPTION
**Pull request by gautierhattenberger**
- based on ChibiOS only (updated to master for proper support of F7)
- requires some specific ram for DMA operations
- SBUS driver updated and support of internal inverted logic option of F7

### CODE REVIEW

File: sw/airborne/mcu_periph/uart.h [hunk 1]
```diff
@@ -33,15 +33,15 @@
...
+#if defined STM32F4 || defined STM32F7 //the F4 and F7 have enough memory
 #define UART_RX_BUFFER_SIZE 256
 #else
 #define UART_RX_BUFFER_SIZE 128
 #endif
 #endif
 
 #ifndef UART_TX_BUFFER_SIZE
-#ifdef STM32F4  //the F4 has enough memory, and the PX4 bootloader needs more then 128
+#if defined STM32F4 || defined STM32F7 //the F4 and F7 have enough memory, and the PX4 bootloader needs more then 128
```
Comment (podhrmic) [inline sw/airborne/mcu_periph/uart.h]: I think @flixr mentioned a while ago that this can be in the board file to avoid yet another ifdef?

File: sw/airborne/modules/loggers/sdlog_chibios/sdLog.c [hunk 2]
```diff
@@ -113,7 +121,7 @@ struct FilePoolUnit {
   uint8_t writeByteSeek;
 };
 
-static  struct FilePoolUnit fileDes[SDLOG_NUM_BUFFER] = {
+static  struct FilePoolUnit IN_DMA_SECTION (fileDes[SDLOG_NUM_BUFFER]) = {
```
Comment (podhrmic) [inline sw/airborne/modules/loggers/sdlog_chibios/sdLog.c]: just making sure, is the logger now backwards compatible with Apogee/PX4?

File: sw/airborne/subsystems/radio_control/sbus_common.c [hunk 3]
```diff
@@ -76,36 +78,31 @@ void sbus_common_init(struct Sbus *sbus_p, struct uart_periph *dev)
...
-    bitInRawBuf++;
-    bitInChannel++;
-
-    if (bitInRawBuf == SBUS_BIT_PER_BYTE) {
-      bitInRawBuf = 0;
-      byteInRawBuf++;
-    }
-    if (bitInChannel == SBUS_BIT_PER_CHANNEL) {
-      bitInChannel = 0;
+  // decode sbus data, unrolling the loop for efficiency
```
Comment (podhrmic) [inline sw/airborne/subsystems/radio_control/sbus_common.c]: why the change here? (unwrapping the loop)?

File: conf/Makefile.chibios [hunk 4]
```diff
@@ -246,9 +251,9 @@ TCSRC =
 TCPPSRC =
 
 # List ASM source files here
-ASMSRC = $(STARTUPASM) $(PORTASM) $(OSALASM)
+ASMXSRC = $(STARTUPASM) $(PORTASM) $(OSALASM)
```
Comment (gautierhattenberger) [inline conf/Makefile.chibios]: I took it from the demos Makefiles, where ASMSRC is void and ASMXSRC is used.

File: conf/boards/chimera_1.0.makefile [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+#
+# Define project name here (target)
+PROJECT = $(TARGET)
+
+# Project specific files and paths (see Makefile.chibios for details)
+CHIBIOS_BOARD_PLATFORM = STM32F7xx/platform.mk
+CHIBIOS_BOARD_PORT = ARMCMx/STM32F7xx/port.mk
+CHIBIOS_BOARD_LINKER = STM32F76xxI.ld
+CHIBIOS_BOARD_STARTUP = startup_stm32f7xx.mk
+ifeq ($(USE_ITCM),1)
```
Comment (gautierhattenberger) [inline conf/boards/chimera_1.0.makefile]: It is a fast RAM that is only accessible wen flashing with Black Magic Probe or with STLINK since dfu-util is not supporting it (can't access the needed address)

File: conf/chibios/chibios_rules.mk [hunk 6]
```diff
@@ -80,8 +84,8 @@ else
   ACSRC += $(CSRC)
   ACPPSRC += $(CPPSRC)
 endif
-ASRC	  = $(ACSRC)$(ACPPSRC)
-TSRC	  = $(TCSRC)$(TCPPSRC)
+ASRC	  = $(ACSRC) $(ACPPSRC)
```
Comment (gautierhattenberger) [inline conf/chibios/chibios_rules.mk]: fixing an old typo

File: sw/airborne/arch/chibios/chconf.h [hunk 7]
```diff
@@ -504,12 +527,30 @@
...
 
 /*===========================================================================*/
 /* Port-specific settings (override port settings defaulted in chcore.h).    */
 /*===========================================================================*/
 
+// point to correct interrupt vector when using ITCM flash on F7
+#if USE_ITCM
+#define CORTEX_VTOR_INIT                    0x00200000U
+#else
+#define CORTEX_VTOR_INIT                    0x08000000U
```
Comment (gautierhattenberger) [inline sw/airborne/arch/chibios/chconf.h]: why not if we have a proper list of options

File: sw/airborne/arch/chibios/mcu_arch.h [hunk 8]
```diff
@@ -44,4 +44,40 @@ extern void mcu_arch_init(void);
 extern bool recovering_from_hard_fault;
 #endif
 
+#include <ch.h>
+
+static inline void mcu_deep_sleep(void)
```
Comment (gautierhattenberger) [inline sw/airborne/arch/chibios/mcu_arch.h]: it is required by the SDLOG system of Chimera and Apogee when closing the files on power down detection

File: sw/airborne/arch/chibios/mcu_periph/gpio_arch.h [hunk 9]
```diff
@@ -30,10 +30,15 @@
 #ifndef GPIO_ARCH_H
 #define GPIO_ARCH_H
 
-#include "hal.h"
+#include <hal.h>
```
Comment (gautierhattenberger) [inline sw/airborne/arch/chibios/mcu_periph/gpio_arch.h]: it is not a system library but still an external library and the `<>` format is used in (most) case for ChibiOS and libopencm3 headers

File: sw/airborne/arch/chibios/mcu_periph/spi_arch.c [hunk 10]
```diff
@@ -140,15 +153,15 @@ static inline uint16_t spi_resolve_slave_pin(uint8_t slave)
...
+ * STM32F4xx and STM32F7xx only)
  * TODO: extend for other architectures too
  *
  * @param[in] t pointer to a @p spi_transaction struct
  */
 static inline uint16_t spi_resolve_CR1(struct spi_transaction *t)
 {
   uint16_t CR1 = 0;
-#if defined(__STM32F10x_H) || defined(__STM32F105xC_H) || defined (__STM32F107xC_H) || defined(__STM32F4xx_H)
+#if defined(__STM32F10x_H) || defined(__STM32F105xC_H) || defined (__STM32F107xC_H) || defined(__STM32F4xx_H) || defined(__STM32F7xx_H)
```
Comment (gautierhattenberger) [inline sw/airborne/arch/chibios/mcu_periph/spi_arch.c]: probably checking for `STM32F1`, `STM32F4` and `STM32F7` is enough

File: sw/airborne/arch/chibios/mcu_periph/uart_arch.c [hunk 11]
```diff
@@ -852,6 +862,14 @@ void uart_periph_set_baudrate(struct uart_periph *p, uint32_t baud )
 void uart_periph_set_mode(struct uart_periph *p __attribute__((unused)), bool tx_enabled __attribute__((unused)),
                           bool rx_enabled __attribute__((unused)), bool hw_flow_control __attribute__((unused))) {}
 
+#if defined STM32F7
+#define __USART_CR1_M USART_CR1_M_0
+#elif defined STM32F4
+#define __USART_CR1_M USART_CR1_M
+#else
+#error unsupported board
```
Comment (gautierhattenberger) [inline sw/airborne/arch/chibios/mcu_periph/uart_arch.c]: I don't use F1 based board, so if you the correct parameters, I'll add them (maybe the same than F4 ?)

File: sw/airborne/arch/chibios/mcu_periph/uart_arch.c [hunk 12]
```diff
@@ -894,6 +912,29 @@ void uart_periph_set_bits_stop_parity(struct uart_periph *p,
   sdStart((SerialDriver*)(p->reg_addr), conf);
 }
 
+#ifdef STM32F7
+/**
+ * Invert data logic
+ */
+void uart_periph_invert_data_logic(struct uart_periph *p, bool invert_rx, bool invert_tx)
```
Comment (gautierhattenberger) [inline sw/airborne/arch/chibios/mcu_periph/uart_arch.c]: yes :)

File: sw/airborne/arch/chibios/subsystems/actuators/actuators_pwm_arch.c [hunk 13]
```diff
@@ -89,40 +89,40 @@ void actuators_pwm_arch_init(void)
...
-  gpio_setup_pin_af(PWM_SERVO_8_GPIO, PWM_SERVO_8_PIN, PWM_SERVO_8_AF);
+  gpio_setup_pin_af(PWM_SERVO_8_GPIO, PWM_SERVO_8_PIN, PWM_SERVO_8_AF, true);
 #endif
 #ifdef PWM_SERVO_9
-  gpio_setup_pin_af(PWM_SERVO_9_GPIO, PWM_SERVO_9_PIN, PWM_SERVO_9_AF);
+  gpio_setup_pin_af(PWM_SERVO_9_GPIO, PWM_SERVO_9_PIN, PWM_SERVO_9_AF, true);
 #endif
 #ifdef PWM_SERVO_10
-  gpio_setup_pin_af(PWM_SERVO_10_GPIO, PWM_SERVO_10_PIN, PWM_SERVO_10_AF);
+  gpio_setup_pin_af(PWM_SERVO_10_GPIO, PWM_SERVO_10_PIN, PWM_SERVO_10_AF, true);
```
Comment (gautierhattenberger) [inline sw/airborne/arch/chibios/subsystems/actuators/actuators_pwm_arch.c]: just for compliance with other arch API. Can be changed in an other pull request

File: sw/airborne/boards/chimera/chibios/v1.0/chimera.h [hunk 14]
```diff
@@ -0,0 +1,560 @@
+#ifndef CONFIG_CHIMERA_1_00_H
+#define CONFIG_CHIMERA_1_00_H
+
+#define BOARD_CHIMERA
+
+/**
+ * ChibiOS board file
+ */
+#include "boards/chimera/chibios/v1.0/board.h"
```
Comment (gautierhattenberger) [inline sw/airborne/boards/chimera/chibios/v1.0/chimera.h]: the board.h file is actually generated, so it is easier to maintain for me

File: sw/airborne/boards/chimera/chibios/v1.0/mcuconf.h [hunk 15]
```diff
@@ -0,0 +1,472 @@
...
+#define STM32_UART_USE_USART6               FALSE /* NO DMA AVAIL */
+#define STM32_UART_USE_UART7                FALSE /* NO DMA AVAIL */
+#define STM32_UART_USE_UART8                FALSE /* NO DMA AVAIL */
+#define STM32_UART_USART1_RX_DMA_STREAM     STM32_DMA_STREAM_ID(2, 5)
+#define STM32_UART_USART1_TX_DMA_STREAM     STM32_DMA_STREAM_ID(2, 7)
+/* #define STM32_UART_USART2_RX_DMA_STREAM     STM32_DMA_STREAM_ID(1, 5) */
+/* #define STM32_UART_USART2_TX_DMA_STREAM     STM32_DMA_STREAM_ID(1, 6) */
+#define STM32_UART_USART3_RX_DMA_STREAM     STM32_DMA_STREAM_ID(1, 1)
+#define STM32_UART_USART3_TX_DMA_STREAM     STM32_DMA_STREAM_ID(1, 3)
+/* #define STM32_UART_UART4_RX_DMA_STREAM      STM32_DMA_STREAM_ID(1, 2) */
```
Comment (gautierhattenberger) [inline sw/airborne/boards/chimera/chibios/v1.0/mcuconf.h]: I can do that

File: sw/airborne/modules/loggers/sdlog_chibios/sdLog.c (hunk 2 above)
Comment (gautierhattenberger) [inline sw/airborne/modules/loggers/sdlog_chibios/sdLog.c]: it should work on both F4 and F7 and only use the special RAM section if needed

File: sw/airborne/subsystems/radio_control/sbus_common.c (hunk 3 above)
Comment (gautierhattenberger) [inline sw/airborne/subsystems/radio_control/sbus_common.c]: as the comment said, it is a bit more efficient like this

File: sw/airborne/mcu_periph/uart.h (hunk 1 above)
Comment (gautierhattenberger) [inline sw/airborne/mcu_periph/uart.h]: I don't mind, but we need to decide if we give priority to need boards (F4, F7, LInux) other older and smaller ones

File: conf/Makefile.chibios (hunk 4 above)
Comment (podhrmic) [inline conf/Makefile.chibios]: I just checked your commit - you are using ChibiOS `master` instead of `stable_16.1.x`
Are you sure you want to switch to the newest (and possibly) buggy ChibiOS? F7xx support seems to be in `stable_16.1.x` too...

Comment (gautierhattenberger) [inline conf/Makefile.chibios]: yep, F7 support is stable branch is buggy, we found a bug that prevent the RTC to start and then the logging is not working correctly (which we consider as a major need when using ChibiOS).

**Comment by podhrmic** [comment]
@gautierhattenberger I see. If you think that the master ChibiOS is stable enough then I am cool with it. Could you pls fix the lia_1.1_chibios makefile with the new path to the linker files so it compiles again? (it is not in the test_conf.xml so thats why travis didn't catch it). Otherwise LGTM.

**Comment by gautierhattenberger** [comment]
which conf is failing ? I tried to compile the `aggiair_rp3_lia` file and it is working (compiling and linking at least).

**Comment by podhrmic** [comment]
conf_tests.xml -> Minion_Lia -> change the board to `lia_1.1_chibios` and then compile (sorry it is not a part of a config already). You ll gt this error:
```
GENERATE /Paparazzi/tmp/paparazzi/var/aircrafts/Minion_Lia/flight_plan.xml from flight_plans/AGGIEAIR/BasicTuning_Launcher.xml
make[1]: Leaving directory '/Paparazzi/tmp/paparazzi'
make[1]: Entering directory '/Paparazzi/tmp/paparazzi'
/Paparazzi/tmp/paparazzi/sw/ext/chibios/os/hal/ports/STM32/STM32F1xx/platform.mk:20: *** missing separator (did you mean TAB instead of 8 spaces?).  Stop.
make[1]: Leaving directory '/Paparazzi/tmp/paparazzi'
Info: module 'gps.xml' has been loaded several times for target ap, merging options
Info: module 'gps.xml' has been loaded several times for target nps, merging options
Makefile.ac:250: recipe for target 'ap.ac_h' failed
make: *** [ap.ac_h] Error 2
```

**Comment by podhrmic** [comment]
@flixr any comments or good to merge?

File: conf/boards/chimera_1.0.makefile (hunk 5 above)
Comment (flixr) [inline conf/boards/chimera_1.0.makefile]: I guess a comment describing this would be nice...

File: conf/boards/apogee_1.0_chibios.makefile [hunk 16]
```diff
@@ -14,9 +14,8 @@ $(TARGET).ARCHDIR = $(ARCH)
 
 RTOS=chibios
 
-# FPU on F4
-USE_FPU=yes
-HARD_FLOAT=yes
```
Comment (flixr) [inline conf/boards/apogee_1.0_chibios.makefile]: what is the reason for not using hard float anymore?

File: sw/airborne/arch/chibios/mcu_arch.h (hunk 8 above)
Comment (flixr) [inline sw/airborne/arch/chibios/mcu_arch.h]: doxygen comment for this function explaining the need and difference to a normal sleep would be nice

File: sw/airborne/arch/chibios/mcu_periph/pwm_input_arch.c [hunk 17]
```diff
@@ -100,14 +100,14 @@ void pwm_input_init(void)
 
 #ifdef USE_PWM_INPUT1
   icuStart(&PWM_INPUT1_ICU, &pwm_input1_cfg);
-  gpio_setup_pin_af(PWM_INPUT1_GPIO_PORT, PWM_INPUT1_GPIO_PIN, PWM_INPUT1_GPIO_AF);
+  gpio_setup_pin_af(PWM_INPUT1_GPIO_PORT, PWM_INPUT1_GPIO_PIN, PWM_INPUT1_GPIO_AF, FALSE);
```
Comment (flixr) [inline sw/airborne/arch/chibios/mcu_periph/pwm_input_arch.c]: should be `false` instead of `FALSE`?

File: sw/airborne/modules/loggers/sdlog_chibios/usbStorage.c [hunk 18]
```diff
@@ -223,11 +223,10 @@ const USBConfig usbConfig = {
 /* Turns on a LED when there is I/O activity on the USB port */
 static void usbActivity(bool active)
 {
-#ifdef GPIOC_LED4
-  palWritePad(GPIOC, GPIOC_LED4, active);
-#else
-  (void)active;
-#endif
+  if (active)
```
Comment (flixr) [inline sw/airborne/modules/loggers/sdlog_chibios/usbStorage.c]: Should this be protected by a `#ifdef SDLOG_USB_LED`?

File: conf/boards/apogee_1.0_chibios.makefile (hunk 16 above)
Comment (gautierhattenberger) [inline conf/boards/apogee_1.0_chibios.makefile]: Compliance with new chibios format.

**Comment by gautierhattenberger** [comment]
We flew with the Chimera this morning and everything went well.
